quic/s2n-quic-core/src/sync/worker.rs

```diff
@@ -0,0 +1,217 @@
+    let state = Arc::new(State::default());
+    let sender = Sender(state.clone());
+    let receiver = Receiver { state, credits: 0 };
+    (sender, receiver)
+}
+
+/// A handle to the receiver side of the worker channel
+///
+/// This handle is used by the worker to wake up when there is work to do.
+pub struct Receiver {
+    state: Arc<State>,
+    credits: usize,
+}
+
+impl Receiver {
+    /// Acquires work to be processed for the Receiver
+    ///
+    /// `None` is returned when there are no more active Senders.
+    #[inline]
+    pub async fn acquire(&mut self) -> Option<usize> {
+        Acquire(self).await
+    }
+
+    /// Polls work to be processed for the receiver
+    ///
+    /// `None` is returned when there are no more active Senders.
+    #[inline]
+    pub fn poll_acquire(&mut self, cx: &mut Context) -> Poll<Option<usize>> {
+        let state = &*self.state;
+
+        macro_rules! acquire {
+            () => {{
+                // take the credits that we've been given by the senders
+                self.credits += state.remaining.swap(0, Ordering::Acquire);
+
+                // if we have any credits then return
+                if self.credits > 0 {
+                    return Poll::Ready(Some(self.credits));
+                }
+            }};
+        }
+
+        // first try to acquire credits
+        acquire!();
+
+        // if we didn't get any credits then register the waker
+        state.receiver.register(cx.waker());
+
+        // make one last effort to acquire credits in case a sender submitted some while we were
+        // registering the waker
+        acquire!();
+
+        // If we're the only ones with a handle to the state then we're done
+        if state.senders.load(Ordering::Acquire) == 0 {
+            return Poll::Ready(None);
+        }
+
+        Poll::Pending
+    }
+
+    /// Marks `count` jobs as finished
+    #[inline]
+    pub fn finish(&mut self, count: usize) {
+        debug_assert!(self.credits >= count);
+        // decrement the number of credits we have
+        self.credits -= count;
+    }
+}
+
+/// A handle to submit work to be done to a worker receiver
+///
+/// Multiple Sender handles can be created with `.clone()`.
+#[derive(Clone)]
+pub struct Sender(Arc<State>);
+
+impl Sender {
+    /// Submits `count` jobs to be executed by the worker receiver
+    #[inline]
+    pub fn submit(&self, count: usize) {
+        let state = &*self.0;
+
+        // increment the work counter
+        state.remaining.fetch_add(count, Ordering::Release);
+
+        // wake up the receiver if possible
+        state.receiver.wake();
+    }
+}
+
+impl Drop for Sender {
+    #[inline]
+    fn drop(&mut self) {
+        let state = &*self.0;
+
+        state.senders.fetch_sub(1, Ordering::Release);
+
+        // wake up the receiver to notify that one of the senders has dropped
+        state.receiver.wake();
+    }
+}
+
+struct State {
+    remaining: CachePadded<AtomicUsize>,
```

Comment:
Did you test with and without `CachePadded` and did it make a noticeable difference in performance?

> Note that N is just a reasonable guess and is not guaranteed to match the actual cache line length of the machine the program is running on.

From the crate, which makes me a bit wary of premature optimization.